cpp/src/binaryop/compiled/util.cpp

```diff
@@ -26,47 +26,60 @@
 namespace cudf::binops::compiled {
 
 namespace {
-/**
- * @brief Functor that returns optional common type of 2 or 3 given types.
- *
- */
+
 struct common_type_functor {
   template <typename TypeLhs, typename TypeRhs>
-  struct nested_common_type_functor {
-    template <typename TypeOut>
-    std::optional<data_type> operator()()
-    {
-      // If common_type exists
-      if constexpr (cudf::has_common_type_v<TypeOut, TypeLhs, TypeRhs>) {
-        using TypeCommon = typename std::common_type<TypeOut, TypeLhs, TypeRhs>::type;
-        return data_type{type_to_id<TypeCommon>()};
-      } else if constexpr (cudf::has_common_type_v<TypeLhs, TypeRhs>) {
-        using TypeCommon = typename std::common_type<TypeLhs, TypeRhs>::type;
-        // Eg. d=t-t
-        return data_type{type_to_id<TypeCommon>()};
-      }
-
-      // A compiler bug may cause a compilation error when using empty initializer list to construct
-      // an std::optional object containing no `data_type` value. Therefore, we should explicitly
-      // return `std::nullopt` instead.
-      return std::nullopt;
+  std::optional<data_type> operator()()
+  {
+    if constexpr (cudf::has_common_type_v<TypeLhs, TypeRhs>) {
+      using TypeCommon = std::common_type_t<TypeLhs, TypeRhs>;
+      return data_type{type_to_id<TypeCommon>()};
     }
-  };
-  template <typename TypeLhs, typename TypeRhs>
-  std::optional<data_type> operator()(data_type out)
+
+    // A compiler bug may cause a compilation error when using empty
+    // initializer list to construct an std::optional object containing no
+    // `data_type` value. Therefore, we explicitly return `std::nullopt`
+    // instead.
+    return std::nullopt;
+  }
+};
+
+struct has_mutable_element_accessor_functor {
```

Comment:
We introduce two new functors and methods to dispatch them (`has_mutable_element_accessor`, `is_constructible`) that we can use at runtime based on the value of `data_type out`. The method `is_constructible` is templated on `ReturnType` because it is known at compile time based on the invoked return type of the binary operator. However, the output type can be used as runtime information with a single dispatch.